### PR TITLE
Added ProtectedToPrivateFixer

### PR DESCRIPTION
**Pull request by Slamdunk**
This fixer converts all attributes and method from protected to private when possible, so if and only if the class:
1. Is final
2. Is not abstract
3. Does not extend any other classes
4. Is not a trait
5. Does not use traits

So only where inheritance is absent

### CODE REVIEW

**Comment by SpacePossum** [comment]
Thanks for working on this!

for ref.: closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1899

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+namespace Symfony\CS\Fixer\Contrib;
+
+use Symfony\CS\AbstractFixer;
+use Symfony\CS\Tokenizer\Token;
+use Symfony\CS\Tokenizer\Tokens;
+
+/**
+ * @author Filippo Tessarotto <zoeslam@gmail.com>
+ */
+class ProtectedToPrivateFixer extends AbstractFixer
```
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: please make the class `final`, we do this on the 2.x/master line for all fixers and also for new fixers on 1.x

File: Symfony/CS/Tests/Fixer/Contrib/ProtectedToPrivateFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ */
+
+namespace Symfony\CS\Tests\Fixer\Contrib;
+
+use Symfony\CS\Tests\Fixer\AbstractFixerTestBase;
+
+/**
+ * @author Filippo Tessarotto <zoeslam@gmail.com>
+ */
+class ProtectedToPrivateFixerTest extends AbstractFixerTestBase
```
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/ProtectedToPrivateFixerTest.php]: please make this test `final` as well and add an `@internal` tag to the class description doc.

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                $prevTokenIndex = $tokens->getPrevMeaningfulToken($prevTokenIndex);
+                $prevToken = $tokens[$prevTokenIndex];
+            }
+
+            if (!$prevToken->isGivenKind(T_PROTECTED)) {
+                continue;
+            }
+
+            $prevToken->clear();
+            $tokens->insertAt(
```
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: maybe you can use `Tokens::overrideAt`?

**Comment by SpacePossum** [comment]
The `use` detection is off I think, lets say we have something like this:
```php
<?php

final class A {
    protected function A() {

    }
}

use \DateTime as Temp;
$d = new Temp();
var_dump($d);
```

**Comment by keradus** [comment]
@SpacePossum 
how?
> I would like to write and add a fixer that removes the final keyword from methods of a final declared class. 

this fixer does not play with `final` keyword removal

**Comment by SpacePossum** [comment]
I think the `protected` for the method can be made `private` in my sample, 
but it won't be fixed because the fixer doesn't check if the `use` is within the `class` body.

**Comment by Slamdunk** [comment]
@SpacePossum your example is covered by the last test case: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2084/files#diff-745abf3775862e89aad00d0c09ce4d33R128

In the meantime I've pushed the changed requested.

**Comment by SpacePossum** [comment]
Thanks for the updates :)
About the sample; I don't think it is. Please try it the other way around, add a test with a class definition first (which should be fixed) and _below_ it an `use` import.

**Comment by Slamdunk** [comment]
You are right: the fixer does not cover any case that has something after the first class.
I'm trying to fix it.

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+            $prevTokenIndex = $tokens->getPrevMeaningfulToken($index);
+            $prevToken = $tokens[$prevTokenIndex];
+
+            if ($prevToken->isGivenKind(T_ABSTRACT)) {
+                return true;
+            }
+
+            $useIndex = $tokens->getNextTokenOfKind($index, array(array(T_USE)));
+            if ($useIndex > 0) {
```
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: @Slamdunk move this check and change it to
```php
if ($useIndex < classOpeningIndex) {
    return false;
}
```

that should do it :)

Comment (Slamdunk) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: Yes, but it's not enough, this fixer should also work for:
```php
class MyClass
{
    protected $var;
}
final class MyClass
{
    protected $var;
}
```
Fixing the latter, but this involves a lot of refactoring of my original code.
I'm working on it, using somthing like [Php4ConstructorFixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.11/Symfony/CS/Fixer/Contrib/Php4ConstructorFixer.php#L30-L34) approach

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+     * @param Tokens $tokens the Tokens instance
+     *
+     * @return bool
+     */
+    private function skipClass(Tokens $tokens, $classIndex)
+    {
+        $prevTokenIndex = $tokens->getPrevMeaningfulToken($classIndex);
+        $prevToken = $tokens[$prevTokenIndex];
+
+        if ($prevToken->isGivenKind(T_ABSTRACT) or !$prevToken->isGivenKind(T_FINAL)) {
```
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: please use `||` over `or` in this case

**Comment by Slamdunk** [comment]
Changes done

File: README.rst [hunk 6]
```diff
@@ -629,6 +629,11 @@ Choose from the list of available fixers:
                         @var should always be written
                         as @type.
 
+* **protected_to_private** [contrib]
+                        Converts all protected
+                        variables and methods to
+                        private if needed.
```
Comment (keradus) [inline README.rst]: it's not needed, it's only possible

**Comment by keradus** [comment]
Could anyone take a look at the fixer, see it's value and make a review ?

**Comment by Slamdunk** [comment]
After rebasing the builds fail because of a [header_comment fix need](https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/154087852#L573) but not only my files fail, also `Symfony/CS/ShutdownFileRemoval.php` is involved and is not part of this PR.

Even if the change is trivial, I am not confortable in touching unrelated files.
Let me know what you want me to do about this.

**Comment by keradus** [comment]
@Slamdunk great catch, I must mess up the merge 1.11->1.12
it's not big change, please apply in this pr, thanks !

**Comment by Slamdunk** [comment]
Done

**Comment by SpacePossum** [comment]
@Slamdunk please consider https://github.com/Slamdunk/PHP-CS-Fixer/pull/1

**Comment by SpacePossum** [comment]
:+1: 
(but lets wait for the test result, btw. please rebase)

**Comment by Slamdunk** [comment]
Everything done

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+     * @param int    $classOpenIndex
+     * @param int    $classCloseIndex
+     */
+    private function fixClass(Tokens $tokens, $classIndex, $classOpenIndex, $classCloseIndex)
+    {
+        for ($index = $classOpenIndex + 1; $index < $classCloseIndex; ++$index) {
+            $prevTokenIndex = $tokens->getPrevMeaningfulToken($index);
+            $prevToken = $tokens[$prevTokenIndex];
+
+            if ($prevToken->isGivenKind(T_STATIC)) {
```
Comment (gharlan) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: why is this extra part needed for `T_STATIC`?
Just skip it like all other tokens apart from `T_PROTECTED`.

**Comment by Slamdunk** [comment]
> Why? Could you give an example for this case, where it would break the behavior? 

https://3v4l.org/6BdYa
In PHP 7 `$v1` raises no error, but `$v2` does

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php (hunk 7 above)
Comment (Slamdunk) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: Yup, legacy copy+paste

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+     * @param int    $classIndex
+     * @param int    $classOpenIndex
+     * @param int    $classCloseIndex
+     *
+     * @return bool
+     */
+    private function skipClass(Tokens $tokens, $classIndex, $classOpenIndex, $classCloseIndex)
+    {
+        $prevToken = $tokens[$tokens->getPrevMeaningfulToken($classIndex)];
+        if ($prevToken->isGivenKind(T_ABSTRACT) || !$prevToken->isGivenKind(T_FINAL)) {
```
Comment (Slamdunk) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: Yup, legacy copy+paste

**Comment by gharlan** [comment]
> https://3v4l.org/6BdYa
> In PHP 7 $v1 raises no error, but $v2 does

Thanks! Didn't know that. 
Theoretically you could change protected methods to private ones anyway in such a case.

**Comment by Slamdunk** [comment]
You are right, but honestly I'd prefer to leave all trait stuff untouched.
Inheritance, both horizontally and vertically, is just a source of mess, even more when changing PHP version.

File: Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php (hunk 7 above)
Comment (Slamdunk) [inline Symfony/CS/Fixer/Contrib/ProtectedToPrivateFixer.php]: No more "outdated diff bla bla", github?

**Comment by keradus** [comment]
Replaced by #2249

**Pull request closed without merging**